src/core/mapping/machine.cc

```diff
@@ -142,9 +142,9 @@ LocalProcessorRange::LocalProcessorRange(uint32_t offset,
 
 const Processor& LocalProcessorRange::operator[](uint32_t idx) const
 {
-  auto local_idx = (idx % total_proc_count_) - offset_;
+  auto local_idx = (idx - offset_) % procs_.size();
```

Comment:
I feel the caller would guarantee that `0 <= idx - offset_ < procs_.size()`, if so we don't need this modulo operator at all. can you confirm if this is really the case?

Reply:
We would actually still need to get modulus. ` idx-offset` can sill be` >=procs_.size()`, the changes in the base mapper just help us to make sure that neighboring indices are mapped to the same proc after we execute the modulus operation

Reply:
if I understand your code correctly, the indices linearized from neighboring points should all be mapped to the same processor id at line 226 by the division with `tasks_per_proc`. so I still don't understand why a processor index minus the offset can still be greater than the number of local processors.

Reply:
yes  you are right, we don't need mod operator

I also found a case when calculated idx is smaller than offset_  (so I added the check):
in cholesky example we have 25 index points distributed between 2 shards with 2 procs per shard.
Here the first 12 points are assigned to shard 0, and the remaining 13 are assigned to shard 1. 
When following the  idx calculation in the mapper, the idx for the point 13 will be `1`, when offset for this idx is `2`. So I added:
```
if (offset_ > idx) local_idx = idx;
```

Reply:
that still doesn't sound right. I think the number of tasks per shard should have been 13 (= (25 + 2 - 1) / 2), then you wouldn't have had that case. keeping the last one getting fewer tasks would likely make the logic cleaner.

Reply:
You were right that we should be dividing by the `tasks_per_proc` , however   I had to modify the proposed formula to deal with the offset per slice. I think the latest version does the right thing now.

Reply:
Note that I still need to do `idx=idx%procs_.size()` in the machine.cc since the logic in the mapper assumes the case when all the tasks are distributed linearly across all shards. However, there are cases when we have all tasks assigned to a single shard (in bincount example we get 2 shards, 4 procs per shard, 5 points in sharding domain where all 5 points are assigned to the same shard)